src/net.c

```diff
@@ -999,6 +1003,110 @@ coap_send(coap_session_t *session, coap_pdu_t *pdu) {
   uint8_t r;
   ssize_t bytes_written;
 
+  if (pdu->code == COAP_RESPONSE_CODE(508)) {
+    /*
+     * Need to prepend our IP identifier to the data as per
+     * https://www.rfc-editor.org/rfc/rfc8768.html#section-4
+     */
+    char addr_str[INET6_ADDRSTRLEN + 8];
+    coap_opt_iterator_t opt_iter;
+    coap_opt_t *opt;
+    size_t hop_limit;
+
+    if (coap_print_addr(&session->addr_info.local, (uint8_t*)addr_str,
+                            INET6_ADDRSTRLEN + 8)) {
+      char *cp;
+      int len;
+
+      if (addr_str[0] == '[') {
+        cp = strchr(addr_str, ']');
+        if (cp) *cp = '\000';
+        if (memcmp(&cp[1], "::ffff:", 7)) {
+          /* IPv4 embedded into IPv6 */
+          cp = &addr_str[8];
+        }
+        else {
+          cp = &addr_str[1];
+        }
+      }
+      else {
+        cp = strchr(addr_str, ':');
+        if (cp) *cp = '\000';
+        cp = addr_str;
+      }
+      len = strlen(cp);
+
+      /* See if Hop Limit option is being used in return path */
+      opt = coap_check_option(pdu, COAP_OPTION_HOP_LIMIT, &opt_iter);
+      if (opt) {
+        coap_option_t result;
+
+        hop_limit =
+          coap_decode_var_bytes (coap_opt_value (opt), coap_opt_length (opt));
+        if (hop_limit == 1) {
+          coap_log(LOG_WARNING, "Proxy loop detected '%s'\n",
+                   (char*)pdu->data);
+          coap_delete_pdu(pdu);
+          return (coap_tid_t)COAP_DROPPED_RESPONSE;
+        }
+        else if (hop_limit < 1 || hop_limit > 255) {
+          /* Something is bad - need to drop this pdu (TODO or delete option) */
+          coap_log(LOG_WARNING, "Proxy return has bad hop limit count '%zd'\n",
+                                 hop_limit);
+          coap_delete_pdu(pdu);
+          return (coap_tid_t)COAP_DROPPED_RESPONSE;
+        }
+        hop_limit--;
+        /* hop_limit is only one byte long, so safe to do */
+        if (coap_opt_parse(opt, 4, &result)) {
+          opt[result.value - opt] = hop_limit;
+        }
+      }
+
+      /* Need to check that we are not seeing this proxy in the return loop */
+      if (pdu->data && opt == NULL) {
+        if (pdu->used_size + 1 <= pdu->max_size) {
+          char *a_match;
+          size_t data_len = pdu->used_size - (pdu->data - pdu->token);
+          pdu->data[data_len] = '\000';
+          a_match = strstr((char*)pdu->data, cp);
+          if (a_match && (a_match == (char*)pdu->data || a_match[-1] == ' ') &&
+              ((size_t)(a_match - (char*)pdu->data + len) == data_len ||
+               a_match[len] == ' ')) {
+            coap_log(LOG_WARNING, "Proxy loop detected '%s'\n",
+                     (char*)pdu->data);
+            coap_delete_pdu(pdu);
+            return (coap_tid_t)COAP_DROPPED_RESPONSE;
+          }
+        }
+      }
+      if (pdu->used_size + len + 1 <= pdu->max_size) {
```

Comment:
It is correct as this is checking that there is enough space to add in the current IP plus space separator..  I have now made coap_pdu_check_resize() internally available and using that instead of coap_pdu_resize().